tianshou/policy/imitation/cql.py

```diff
@@ -208,12 +208,21 @@ def learn(self, batch: Batch, **kwargs: Any) -> Dict[str, float]:
         random_actions = torch.FloatTensor(
             batch_size * self.num_repeat_actions, act.shape[-1]
         ).uniform_(-self.min_action, self.max_action).to(self.device)
-        tmp_obs = obs.unsqueeze(1) \
-            .repeat(1, self.num_repeat_actions, 1) \
-            .view(batch_size * self.num_repeat_actions, obs.shape[-1])
-        tmp_obs_next = obs_next.unsqueeze(1) \
-            .repeat(1, self.num_repeat_actions, 1) \
-            .view(batch_size * self.num_repeat_actions, obs.shape[-1])
+
+        if len(obs.shape) > 2:
+            tmp_obs = obs.unsqueeze(1) \
+                .repeat(1, self.num_repeat_actions, 1, 1) \
+                .view(batch_size * self.num_repeat_actions, obs.shape[-2], obs.shape[-1])
+            tmp_obs_next = obs_next.unsqueeze(1) \
+                .repeat(1, self.num_repeat_actions, 1, 1) \
+                .view(batch_size * self.num_repeat_actions, obs.shape[-2], obs.shape[-1])
+        else:
+            tmp_obs = obs.unsqueeze(1) \
+                .repeat(1, self.num_repeat_actions, 1) \
+                .view(batch_size * self.num_repeat_actions, obs.shape[-1])
+            tmp_obs_next = obs_next.unsqueeze(1) \
+                .repeat(1, self.num_repeat_actions, 1) \
+                .view(batch_size * self.num_repeat_actions, obs.shape[-1])
```

Comment:
Is there a more general way to write this part of code?

Reply:
It is so CQl would take into account stacked observation for RNNs. Not sure if it can be done in a another way. But with this modification the code runs correctly for CQL with RNNs. However there is still the general problem that algorithms don't learn with RNNs as they are implemented currently in tianshou.